### PR TITLE
sqld: Add `BatchLogEntries` and replication rpc auth

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR adds a new rpc endpoint that doesn't stream log entries fit for libsql and additionally adds auth to the replication service rpc code which is only enabled for the public facing http server and not the internal one.

### CODE REVIEW

File: sqld/proto/replication_log.proto [hunk 1]
```diff
@@ -20,8 +20,13 @@ message Frame {
...
 }
 
+message Frames {
+    repeated Frame frames = 1;
+}
+
 service ReplicationLog {
     rpc Hello(HelloRequest) returns (HelloResponse) {}
     rpc LogEntries(LogOffset) returns (stream Frame) {}
+    rpc LogEntriesSnapshot(LogOffset) returns (Frames) {}
```
Comment (haaawk) [inline sqld/proto/replication_log.proto:30]: I'm not sure if Snapshot is a good name since we're using that name already for different concept.

Comment (haaawk) [reply sqld/proto/replication_log.proto:30]: Maybe something like LogEntriesNonblocking would better describe the difference from LogEntries?

File: sqld/src/rpc/replication_log.rs [hunk 2]
```diff
@@ -94,7 +95,7 @@ impl<S: futures::stream::Stream + Unpin> futures::stream::Stream for StreamGuard
         self: std::pin::Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Self::Item>> {
-        self.get_mut().s.poll_next_unpin(cx)
+        Pin::new(&mut self.get_mut().s).poll_next(cx)
```
Comment (haaawk) [inline sqld/src/rpc/replication_log.rs:111]: Out of curiosity, what is this change doing?

File: sqld/src/rpc/replication_log.rs [hunk 3]
```diff
@@ -34,11 +36,22 @@ impl ReplicationLogService {
...
+            auth,
+        }
+    }
+
+    fn authenticate<T>(&self, req: &tonic::Request<T>) -> Result<(), Status> {
+        if let Some(auth) = &self.auth {
+            let _ = auth.authenticate_grpc(req)?;
+            Ok(())
+        } else {
+            Ok(())
```
Comment (haaawk) [inline sqld/src/rpc/replication_log.rs:54]: Ultra nit: maybe:
```
if let Some(auth) = &self.auth {
  let _ = auth.authenticate_grpc(req)?;
}
Ok(())
```
?

**Review by haaawk — APPROVED** [review]

File: sqld/proto/replication_log.proto (hunk 1 above)
Comment (haaawk) [reply sqld/proto/replication_log.proto:30]: or maybe LogEntriesBatch?

File: sqld/src/rpc/replication_log.rs (hunk 2 above)
Comment (LucioFranco) [reply sqld/src/rpc/replication_log.rs:111]: I removed the `futures` crate and moved to tokio_stream (which is better but happy to elaborate on slack). The only issue is that tokio_stream's ext trait doesn't have the unpin and boxed fn so I replaced those usages. This just does the same thing as the fn does but done manually.

File: sqld/src/replication/primary/frame_stream.rs [hunk 4]
```diff
@@ -84,6 +90,12 @@ impl Stream for FrameStream {
                 }
 
                 Err(LogReadError::Ahead) => {
+                    // If we don't wait to wait for more then lets end this stream
+                    // without subscribing for more frames
+                    if !self.wait_for_more {
+                        return Poll::Ready(None);
```
[the remark this answers is not in the document]
Comment (LucioFranco) [reply sqld/src/replication/primary/frame_stream.rs:96]: Yup, good catch, Ive added that as a new line.

**Review by psarna — APPROVED** [review]

File: sqld/src/rpc/replication_log.rs (hunk 2 above)
Comment (haaawk) [reply sqld/src/rpc/replication_log.rs:110]: Cool. No objections. I was just curious. Thanks

**Pull request merged**